rfcs/0001-messages.md

```diff
@@ -0,0 +1,254 @@
+
+```rust
+impl Validate for Header {
+    fn validate(&self) -> &Self {
+        if self.sender.0 == self.receiver.0 {
+            panic!("sender and receiver are the same");
+        }
+        self
+    }
+}
+```
+
+Then to create a valid `Header` we call:
+
+```rust
+let header = Validate::validate(&Header {
+    ..
+}).clone();
+```
+
+The receiver side will validate the header as:
+
+```rust
+msg.header.validate();
+```
+
+## Serialization/Deserialization
+
+Each message struct needs to serialize to bytes representation before it is sent through the wire and must deserialize to the same struct (round trip) on the receiver side. We use `serde` and macro derivations (`Serialize` and `Deserialize`) to automatically implement where possible.
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u8
+1     | receiver   | u8
+
+### Primitive types
+
+`Payload`s use data types that we need to specify first. We have 2 primitive types inside the payload messages:
+
+#### `Scalar`
+
+`Scalar` is a an alias for `jubjub::Fr` and this is a `[u64; 4]` as documented in https://github.com/zkcrypto/jubjub/blob/main/src/fr.rs#L16
+
+#### `AffinePoint`
+
+Much of the math in FROST is done using `jubjub::ExtendedPoint`. This is a structure with 5 `jubjub::Fq`s as defined in https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L128-L134
+
+Each `Fq` needed to form a `jubjub::ExtendedPoint` are `Scalar`s of `bls12_381` crate. Scalar here is `[u64; 4]` as documented in https://github.com/zkcrypto/bls12_381/blob/main/src/scalar.rs#L16
+
+For message exchange `jubjub::AffinePoint`s are a better choice as they are shorter in bytes, they are formed of 2 `jubjub::Fq` instead of 5: https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L70-L73
+
+Conversion from one type to the other is trivial:
+
+https://docs.rs/jubjub/0.6.0/jubjub/struct.AffinePoint.html#impl-From%3CExtendedPoint%3E
+https://docs.rs/jubjub/0.6.0/jubjub/struct.ExtendedPoint.html#impl-From%3CAffinePoint%3E
+
+### Payload
+
+Payload part of the message is variable in size and depends on message type.
+
+#### `MsgDealerBroadcast`
+
+Bytes  | Field name  | Data type
+-------|-------------|-----------
+256    | secret_key  | Scalar
+512    | commitments | AffinePoint
+256+32 | group_public| (AffinePoint, [u8; 32])
+
+#### `MsgCommitments`
+
+Bytes | Field name | Data type
+------|------------|-----------
+512   | hiding     | AffinePoint
+512   | binding    | AffinePoint
+
+#### `MsgSigningPackage`
+
+Bytes      | Field name     | Data type
+-----------|----------------|-----------
+?          | message        | [u8]
```

Comment:
We don't know the size of the message. We can either limit the amount of bytes or move the `message` to the end of the structure and use `read_to_end()` or similar. Suggestions?